drivers/source/usb/USBDevice.cpp

```diff
@@ -851,8 +851,8 @@ void USBDevice::ep0_setup()
 {
     assert_locked();
 
+    MBED_ASSERT(_device.state < Default);
```

Comment:
This assert, like many others, is backwards - should be `>= Default`.

But if you change it to that, I can imagine the compiler still complaining - you still can't reach the `return`.

There's a logical problem in the code here - an `assert` is an assertion that the condition cannot happen. If it can't happen, you don't need the `if`. The non-release build checks the assertion with the assert, and the release build no longer needs any checks.

On code size grounds, I would be in favour of leaving the asserts in, and dropping the code trying to handle the assertion failure manually:

    MBED_ASSERT(_device.state >= Default);

    _setup_ready = true;

Alternatively, you could conceivably do it all manually, but it would be fat in release builds:

     #if MBED_TRAP_ERRORS // or whatever it is
     MBED_ERROR(xxx);
     #else
     return;
     #endif

But I think that's not productive.

Just asserting would match `assert_locked()` - that just asserts, but has no exit path. (Although I note that is still inefficient for release builds, as it's a `virtual` function not in the header. It should really be inlined and non-virtual so it can be eliminated).

Reply:
With the release build the function returns if `_device.state < Default` I want to keep that behavior as I am not familiar with this code. With your first suggestion the function will not return if `_device.state < Default`.

My goal is to keep the exact same behavior whilst getting rid of the warning. Your suggestion change the behavior. This can be done in a separate PR.

Reply:
Would this actually remove the warnings with the assert the correct way round though? The condition is currently backwards.

> My goal is to keep the exact same behavior whilst getting rid of the warning.

I'm always wary of that being a goal. Warnings, unless spurious, are something that need to be properly resolved, not just hidden. They can be indicating that the behaviour is wrong, and we don't want to leave the incorrect behaviour while taking away the thing that was pointing it out. Silencing warnings should never be the end in itself. (If it was, then just putting `-W` on the compiler would be the easiest technique).

If anyone is relying on current behaviour, they can only be relying on it for release builds - the develop or debug builds would be asserting and halting. Seems a bit unlikely, but I guess it's possible. (A comparator would be the extra mutex-in-IRQ checks that used to be done only in debug builds).

Is there a reason to be especially wary about changing behaviour? Are we trying to land warning removals in a patch release? I'm not sure we should be doing that anyway, even if we think we're not changing behaviour. I'd be inclined just to do one correct fix for minor release.

@JanneKiiskila